rio/src/RIO/File.hs

```diff
@@ -0,0 +1,345 @@
+         fileFd <- handleToFd fileHandler
+         fsyncDescriptor "File" (FD.fdFD fileFd)
+           `finally` hClose fileHandler
+         renameFile tmpFp fp
+         fsyncDescriptor "Directory" dirFd)
+    closeDirectory
+  where
+    tmpFilePath = toTmpFilePath filePath
+    fsyncDescriptor name fd =
+      void $
+      throwErrnoIfMinus1 ("closeFileDurableAtomic - fsync(" <> name <> ")") $
+      c_safe_fsync fd
+    renameFile tmpFp origFp =
+      void $
+      throwErrnoIfMinus1Retry "closeFileDurableAtomic - renameFile" $
+      c_safe_renameat dirFd tmpFp dirFd origFp
+    closeDirectory =
+      void $
+      throwErrnoIfMinus1Retry "closeFileDurableAtomic - close(Directory)" $
+      c_close dirFd
+
+#endif
+
+-- | After a file is closed, it opens it again and executes fsync internally on
+-- both the file and the directory that contains it.
+--
+-- [The effectiveness of calling this function is
+-- debatable](https://stackoverflow.com/questions/37288453/calling-fsync2-after-close2/50158433#50158433),
+-- as this relies on internal implementation details at the Kernel level that
+-- might change. We argue that, despite this fact, calling this function may
+-- bring benefits in terms of durability
+--
+-- @since 0.1.6
+ensureFileDurable :: MonadUnliftIO m => FilePath -> m ()
+ensureFileDurable absFp =
+#if WINDOWS
+  absFp `seq` return ()
+#else
+  bracket (openFileDurable absFp ReadMode)
+          (uncurry closeFileDurable)
+          (const $ return ())
+#endif
+
+
+-- | Similar to 'writeFileBinary', but it also ensures that changes executed to
+-- the file are guaranteed to be durable. It internally uses fsync and makes
+-- sure it synchronizes the file on disk.
+--
+-- @since 0.1.6
+writeFileBinaryDurable :: MonadUnliftIO m => FilePath -> ByteString -> m ()
+writeFileBinaryDurable absFp bytes =
+#if WINDOWS
+  writeFileBinary absFp bytes
+#else
+  withFileDurable absFp WriteMode (liftIO . (`hPut` bytes))
+#endif
+
+-- | Similar to 'writeFileBinary', but it also guarantes that changes executed
+-- to the file are durable, also, in case of failure, the modified file is never
+-- going to get corrupted. It internally uses fsync and makes sure it
+-- synchronizes the file on disk.
+--
+-- @since 0.1.6
+writeFileBinaryDurableAtomic :: MonadUnliftIO m => FilePath -> ByteString -> m ()
+writeFileBinaryDurableAtomic fp bytes =
+#if WINDOWS
+  writeFileBinary fp bytes
+#else
+  withFileDurableAtomic fp WriteMode (liftIO . (`hPut` bytes))
+#endif
+
+-- | Opens a file with the following guarantees:
+--
+-- * Successfully closes the file in case of an asynchronous exception
+--
+-- * It reliably saves the file in the correct directory; including edge case
+--   situations like a different device being mounted to the current directory,
+--   or the current directory being renamed to some other name while the file is
+--   being used.
+--
+-- * It ensures durability by executing fsync before close
+--
+-- @since 0.1.6
+withFileDurable ::
```

Comment:
Should this instead be `withFileBinaryDurable`? We tend to prefer binary file handles everywhere else.

Reply:
These names are getting pretty _Germanic_ 🙈

Reply:
Ugh, I messed up, it _should_ be `withBinaryFileDurable` for consistency, sorry.